### PR TITLE
RMI-192 Code Coverage Library/Plugin

### PR DESCRIPTION
**Pull request by juancho088**
- New base code
- All tests scenarios
- Coverage implementation
- Sonarqube integration
- Github integration

### CODE REVIEW

File: src/com/travelstart/plugins/jenkins/sonar/Coverage.groovy [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+            else {
+                throw new DataIntegrityException(body.toString())
+            }
+        }
+
+        return result
+    }
+
+    @Override
+    def update(final String prId, final String state, final String targetUrl, final String description) {
```
Comment (ashleighj) [inline src/com/travelstart/plugins/jenkins/sonar/Coverage.groovy:91]: Won't this always return a Map? Can we type it? (Same for line 24)

File: vars/compareCoverageMetric.groovy [hunk 2]
```diff
@@ -0,0 +1,34 @@
+#!/usr/bin/env groovy
+import com.travelstart.plugins.exceptions.PluginException
+import com.travelstart.plugins.jenkins.sonar.Coverage
+
+def call(final Map args) {
+
+    withCredentials([string(credentialsId: 'github-sonarqube-oauth', variable: 'GIT_SONAR_TOKEN_LOCAL')]) {
+        final String gitToken = env.GIT_SONAR_TOKEN ?: "${GIT_SONAR_TOKEN_LOCAL}"
+
+        final String gitRepo = env.GIT_REPO ?: sh(returnStdout: true, script: 'git config remote.origin.url').trim().replace("https://github.com","").replace(".git", "")
```
Comment (ashleighj) [inline vars/compareCoverageMetric.groovy]: Maybe we should have uniformity ito double quotes vs single quotes for strings?

File: test/src/com/travelstart/plugins/jenkins/sonar/MetricTest.groovy [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+        setupServer()
+    }
+
+    void cleanup() {
+        cleanupServer()
+    }
+
+    def "Raise An Exception if HTTP response status code is not between 200 and 299"() {
+        given:
+            def response = mock(HttpURLConnection.class)
```
Comment (ashleighj) [inline test/src/com/travelstart/plugins/jenkins/sonar/MetricTest.groovy]: Can just write `HttpURLConnection` without the `.class` in Groovy

File: test/src/com/travelstart/plugins/jenkins/sonar/MetricTest.groovy [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+            def expectedResponse = importFile("404.json").text
+
+            assertThat(e.code, equalTo(404))
+            assertThat(e.message, equalTo("Not Found"))
+            assertThat(e.body, equalTo(expectedResponse))
+    }
+
+    def "Continue without failure if HTTP response code is between 200 and 299"() {
+        given:
+            def response = mock(HttpURLConnection.class)
```
Comment (ashleighj) [inline test/src/com/travelstart/plugins/jenkins/sonar/MetricTest.groovy]: Same

File: src/com/travelstart/plugins/jenkins/sonar/Coverage.groovy (hunk 1 above)
Comment (juancho088) [reply src/com/travelstart/plugins/jenkins/sonar/Coverage.groovy:91]: according to JsonSlurper sometimes could be a Map and sometimes a List
```
 a data structure of lists and maps
```
so in general they return object ```public Object parseText(String text)``` so I don't want to put Object (that will ruin the dynamic typing)

**Pull request merged**